File: Data Structures & Algorithms/binary-tree-diameter/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func diameterOfBinaryTree(root *TreeNode) int {
	var res int
	var dfs func(*TreeNode) int
	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		} else {
			right := dfs(root.Right)
			left :=  dfs(root.Left)
			res = max(res, right + left)
			return 1 + max(left, right)
		}
	}
	dfs(root)

	return res
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    dfs(root){
        if (!root) {
            return [1,0]
        }

        let left = this.dfs(root.left)
        let right = this.dfs(root.right)
        const balanced =  (left[0] === 1 && right[0] === 1 && Math.abs(left[1] - right[1]) <= 1)  
        
        return [balanced ? 1 : 0, 1 + Math.max(left[1], right[1])]
    }
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
       return this.dfs(root)[0] === 1
    }
     
}
